File: src/angel/agents/issue_review/state.py
import operator
from typing import Annotated, NamedTuple, TypedDict

from angel.agents._shared import ItemFailure, Review
from angel.clients.github import Comment, Issue


class IssueContext(NamedTuple):
    issue: Issue
    comments: list[Comment]


class IssueReviewState(TypedDict):
    issues: list[IssueContext]
    reviews: list[Review]
    failures: Annotated[list[ItemFailure], operator.add]

File: src/agents/issue-review/state.ts
import { Annotation } from "@langchain/langgraph";
import type { Comment, Issue } from "../../clients/github.js";
import type { ItemFailure, Review } from "../shared.js";

export interface IssueContext {
  readonly issue: Issue;
  readonly comments: Comment[];
}

function replace<T>(_current: T, next: T): T {
  return next;
}

export const IssueReviewAnnotation = Annotation.Root({
  issues: Annotation<IssueContext[]>({ reducer: replace, default: () => [] }),
  reviews: Annotation<Review[]>({ reducer: replace, default: () => [] }),
  failures: Annotation<ItemFailure[]>({
    reducer: (current, next) => [...current, ...next],
    default: () => [],
  }),
});

export type IssueReviewState = typeof IssueReviewAnnotation.State;
